File: src/content/docs/guides/fact/fact.lua
local function factorial(n)
    if n == 0 or n == 1 then return 1 end

    return n * factorial(n - 1)
end

local function is_integer(n) return type(n) == "number" and math.floor(n) == n end

print("enter a number:")
local n = assert(io.read("*number"), "your input is not a valid number")

if n < 0 or not is_integer(n) then
    error("your input is not a non-negative number");
end

print(factorial(n))

File: src/content/docs/guides/fact/fact.ts
import readline from 'node:readline';

function factorial(n: number): number {
  let result = 1;
  for (let i = 1; i <= n; i++) {
    result *= i;
  }
  return result;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

new Promise<string>((resolve) => rl.question('enter a number:\n', resolve))
  .then((raw) => {
    const n = parseInt(raw);

    if (isNaN(n)) {
      throw Error('your input is not a valid number');
    }
    if (n < 0) {
      throw Error('your input is not a non-negative number');
    }

    return factorial(n);
  })
  .then(console.log)
  .catch(console.error)
  .finally(() => {
    rl.close();
  });
